proposals/interface-types/working-notes/Types.md

```diff
@@ -37,36 +54,157 @@ Integer values.
+>Note: In a future revision, this may be reinterpreted as a macro of a more
+>explicit designation of the `sequence` type.
+
+Note that there is no implied representation or encoding implied in this.
 
-A `record` type is valid iff:
- * `|$fields| > 0`
- * For every `field` in `$fields`
-  * `#interface-valtype` is valid
 
-## `array`
+### Structured Value Types
 
-The `array` interface value type is a compound type defined by the following abstract syntax:
+Structured types allow the representation of more complex combinations of
+values. There are five forms of structured type: arrays, sequences, records,
+protocols and algebraic variants.
 
 ```
-#array ::= array #interface-valtype
+structuretype ::= optiontype | eithertype |
+   arraytype | sequencetype | recordtype | varianttype
+```
+
+#### Option Type
+
+Option types are used to model nullability in types.
+
+```
+optiontype ::= option type
+```
+
+The `option` type can be viewed as a use of type variants -- with one variant
+being the enumerated symbol `$none` and the other being the single value wrapped
+in a `$some` variant.
+
+#### Either Type
+
+The `either` type is used to model situations where one of two values may be
+returned by a function. A classic case for this is in modeling error returns in
+APIs.
+
+```
+eithertype ::= either type
+```
+
+#### Arrays
+
+An `array` consists of a finite vector of elements, each of which has the same
+type.
+
+```
+arraytype ::= array type
+```
+
+Both the type and the representation of each element of the array is assumed to
+be identical. Furthermore, arrays are assumed to have a know element
+count. (Indexing into arrays is not part of this specification.)
+
+#### Sequences
+
+A sequence consists of an ordered collection of elements; each of which has the
+same type.
+
+Sequence elements may not have identical representation; nor is it necessarily
+known how many elements there are in a sequence.
+
+```
+sequencetype ::= sequence type
+```
+
+#### Records
+
+A `record` type is a tuple composition of types. 
+
 ```
+#record ::= record [ vec(#interfacetype) ]
+```
+
+Fields in a record are accessed via their index; for convenience, field indices
+may be given a local name:
+
+```
+record [ $name string $age u64]
+```
+
+#### Variant Types
+
+A `variant` consists of an ordered collection of alternative vectors of types.
+
+```
+varianttype ::= oneof [ vec(vec(#interfacetype)) ]
+```
+
+Variants are accessed by their index in the vector of alternatives. For
+convenience, variants may be given a local name:
+
+```
+variant [ $id [s64] $name [string] ]
+```
+
+A given variant may not have any type arguments; in which case it corresponds to
+an enumerated symbol. For example, the boolean type may be viewed as a synonym
+for:
+
+```
+(type $boolean (variant ($false) ($true)))
+```
+
+Additionally, the option type may be viewed as a synonym for:
```

Comment:
Do we need option and variant as top-level types if we have variant?

Reply:
I believe so. If you unpack why we have boolean called out, the same reasoning applies to option type (nullability is a key use case).

Reply:
Right, but we don't have `integraltype ::= ... | boolean`, but we do have `structuretype ::= ... | optiontype`

It makes perfect sense to show boolean and option as called-out examples, because of how common they are, I'm just wondering if optiontype needs to exist explicitly.

Although by showing how we can think of it as being equivalent to a variant, I don't think it matters much either way.

Reply:
Ah, I think I understand now. Maybe make variant type a top level and have option/either as additional cases? Or having a special macro section? (I think I prefer the latter)

Reply:
I agree that if we're having `either` and `option` as 'macros' that explicitly expand to `variant` we should have them in their own section.